pkg/reconciler/apis/apibinding/apibinding_permission_claim_reconcile.go

```diff
@@ -0,0 +1,200 @@
+
+	if len(apiBinding.Spec.AcceptedPermissionClaims) == 0 && len(apiBinding.Status.ObservedAcceptedPermissionClaims) == 0 {
+		return nil
+	}
+
+	lc := logicalcluster.From(apiBinding)
+
+	dc := c.discoveryClusterClient.WithCluster(lc)
+
+	// TODO: this is not scalable and we will need to figure out a better way.
+	// One option is an informer/cache
+	apiGroupResources, err := restmapper.GetAPIGroupResources(dc)
+	if err != nil {
+		return err
+	}
+	mapper := restmapper.NewDiscoveryRESTMapper(apiGroupResources)
+
+	// get the added permission claims if added, add are just requeue the object.
+	addedPermissionClaims := map[schema.GroupVersionResource][]permissionClaim{}
+	for _, acceptedPC := range apiBinding.Spec.AcceptedPermissionClaims {
+		found := false
+		for _, observedPC := range apiBinding.Status.ObservedAcceptedPermissionClaims {
+			if reflect.DeepEqual(acceptedPC, observedPC) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			gvr, err := mapper.ResourcesFor(schema.GroupVersionResource{Group: acceptedPC.Group, Resource: acceptedPC.Resource})
+
+			if err != nil {
+				return err
+			}
+			if len(gvr) == 0 {
+				return fmt.Errorf("unable to find the version for resource %v", acceptedPC.GroupResource)
+			}
+			key, label, err := permissionclaims.PermissionClaimToLabel(acceptedPC)
+			if err != nil {
+				return err
+			}
+
+			claims := addedPermissionClaims[gvr[0]]
+			if claims == nil {
+				claims = []permissionClaim{}
+			}
+			claims = append(claims, permissionClaim{
+				claim: acceptedPC,
+				label: label,
+				key:   key,
+			})
+			addedPermissionClaims[gvr[0]] = claims
+		}
+	}
+
+	removedPermissionClaims := map[schema.GroupVersionResource][]permissionClaim{}
+	for _, observedPC := range apiBinding.Status.ObservedAcceptedPermissionClaims {
+		found := false
+		for _, acceptedPC := range apiBinding.Spec.AcceptedPermissionClaims {
+			if reflect.DeepEqual(observedPC, acceptedPC) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			gvr, err := mapper.ResourcesFor(schema.GroupVersionResource{Group: observedPC.Group, Resource: observedPC.Resource})
+			if err != nil {
+				return err
+			}
+			if len(gvr) == 0 {
+				return fmt.Errorf("unable to find the version for resource %v", observedPC.GroupResource)
+			}
+			key, label, err := permissionclaims.PermissionClaimToLabel(observedPC)
+			if err != nil {
+				return err
+			}
+			claims := removedPermissionClaims[gvr[0]]
+			if claims == nil {
+				claims = []permissionClaim{}
+			}
+			claims = append(claims, permissionClaim{
+				claim: observedPC,
+				label: label,
+				key:   key,
+			})
+			removedPermissionClaims[gvr[0]] = claims
+		}
+	}
+
+	klog.V(2).Infof("permission controller adding permission claims: %#v, removing: %#v", addedPermissionClaims, removedPermissionClaims)
+
+	for gvr, claims := range addedPermissionClaims {
+		objs, err := c.dynamicClusterClient.
+			Cluster(lc).
+			Resource(gvr).
+			List(ctx, v1.ListOptions{})
```

Comment:
this won't scale, we can't do this.

Alternative to check out: we have the dynamic informer used for e.g. the resource controller. Grep for ddsif in kcp. It's not wired up in pkg/server/server.go and can be consumed by other informers (wasn't like that a week ago). It has byWorkspace indexes. So you can loop of informers and then by workspace to get all objects. You will also not need GVRs in advance. You get the GVRs from the ddsif.

Note: the ddsif only give you partial meta data. But that's enough here.

Reply:
Adding test.